src/client/datascience/notebook/intellisense/logReplayService.ts

```diff
@@ -0,0 +1,287 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+'use strict';
+import { injectable, inject } from 'inversify';
+import * as os from 'os';
+import * as vscode from 'vscode';
+import { IExtensionSingleActivationService } from '../../../activation/types';
+import { IApplicationShell, ICommandManager } from '../../../common/application/types';
+import { IFileSystem } from '../../../common/platform/types';
+import { IConfigurationService, IDisposableRegistry } from '../../../common/types';
+import { Commands, EditorContexts } from '../../constants';
+import * as lspConcat from '@vscode/lsp-notebook-concat';
+import { NOTEBOOK_SELECTOR, PYTHON_LANGUAGE } from '../../../common/constants';
+import * as protocol from 'vscode-languageserver-protocol';
+import { traceInfo } from '../../../common/logger';
+import { ContextKey } from '../../../common/contextKey';
+import { sleep, waitForCondition } from '../../../common/utils/async';
+
+/**
+ * Class used to replay pylance log output to regenerate a series of edits. SHIFT+F10 will play the steps one at a time
+ */
+@injectable()
+export class LogReplayService implements IExtensionSingleActivationService {
+    private steps: protocol.DidChangeTextDocumentParams[] = [];
+    private index = -1;
+    private converter: lspConcat.NotebookConverter | undefined;
+    private activeNotebook: vscode.NotebookDocument | undefined;
+    private isLogActive: ContextKey | undefined;
+    constructor(
+        @inject(ICommandManager) private readonly commandService: ICommandManager,
+        @inject(IDisposableRegistry) private readonly disposableRegistry: IDisposableRegistry,
+        @inject(IApplicationShell) private readonly appShell: IApplicationShell,
+        @inject(IFileSystem) private readonly fs: IFileSystem,
+        @inject(IConfigurationService) private readonly configService: IConfigurationService
+    ) {}
+    public async activate(): Promise<void> {
+        this.disposableRegistry.push(
+            this.commandService.registerCommand(Commands.ReplayPylanceLog, this.replayPylanceLog, this)
+        );
+        this.disposableRegistry.push(
+            this.commandService.registerCommand(Commands.ReplayPylanceLogStep, this.step, this)
+        );
+        this.isLogActive = new ContextKey(EditorContexts.ReplayLogLoaded, this.commandService);
+        void this.isLogActive.set(false);
+    }
+
+    private async replayPylanceLog() {
+        if (vscode.window.activeNotebookEditor) {
+            const file = await this.appShell.showOpenDialog({ title: 'Open Pylance Output Log' });
```

Comment:
Silly question, but I'm assuming this means we can take the raw output from the output panel.
I.e. I can just copy the output and paste into a file, correct?

Reply:
Yep. That's how it works.

Reply:
Is there something we can do to ensure the format of the logs don't change without updating this code?
I'm concerned that once we have this stuff working for a while, then when we do want to use it, the format will be out of sync.

Reply:
Doesn't matter, its just for us (dev tool).